### PR TITLE
#175070987 account overview endpoint

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Add account overview endpoint

#### Description of Task to be completed?
Add endpoint to return financial overview of user account 

#### How should this be manually tested?
`GET /api/v1/user/overview`

#### What are the relevant pivotal tracker stories?
#175070987

### CODE REVIEW

File: server/services/user.service.js [hunk 1]
```diff
@@ -328,3 +330,17 @@ export const removePropertyFromFavorites = async (favorite) => {
...
 };
+
+export const getAccountOverview = async (userId) => {
+  let contributionReward = await calculateContributionReward(userId);
+  contributionReward = contributionReward.length > 0 ? contributionReward[0].contributionReward : 0;
+
+  let totalAmountPaid = await getTotalAmountPaidByUser(userId);
+  totalAmountPaid = totalAmountPaid.length > 0 ? totalAmountPaid[0].totalAmountPaid : 0;
+
+  const overview = {
```
Comment (andela-hpopoola) [inline server/services/user.service.js:341]: This serves no purpose, we can return the whole object.

File: server/services/user.service.js [hunk 2]
```diff
@@ -328,3 +330,17 @@ export const removePropertyFromFavorites = async (favorite) => {
...
+    contributionReward,
+    totalAmountPaid,
+  };
+  return overview;
```
Comment (andela-hpopoola) [inline server/services/user.service.js:345]:
```suggestion
  return { 
     contributionReward,
    totalAmountPaid,
  };
```

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. 

I thought we agreed that we should add the referral rewards too to this and use test to verify that it works accordingly? Please lets add this so that we can wrap this up. Thank you.

File: server/routes/user.routes.js [hunk 3]
```diff
@@ -430,4 +430,25 @@ router.post(
   UserController.removePropertyFromFavorites,
 );
 
+/**
+ * @swagger
+ * path:
+ *  /user/overview:
```
Comment (andela-hpopoola) [inline server/routes/user.routes.js:436]: Overview is too generic, lets make this to be account-overview so that it can be explicit.
```suggestion
 *  /user/account-overview:
```

File: server/controllers/user.controllers.js [hunk 4]
```diff
@@ -158,6 +159,18 @@ const UserController = {
...
       .catch((error) => next(error));
   },
+
+  getAccountOverview(req, res, next) {
+    const userId = req.user._id;
+    getAccountOverview(userId)
+      .then((overview) => {
+        res.status(httpStatus.OK).json({
+          success: true,
+          overview,
```
Comment (andela-hpopoola) [inline server/controllers/user.controllers.js:169]: I prefer we should be more explicit to make the endpoint easier to work with.
```suggestion
          accountOverview,
```

File: server/routes/user.routes.js [hunk 5]
```diff
@@ -430,4 +430,25 @@ router.post(
...
+ *          description: the auto generated user token via jwt
+ *      summary: Returns financial overview of user (contribution reward, referral bonus etc)
+ *      tags: [User]
+ *      responses:
+ *        '200':
+ *          description: Returns overview
+ *        '500':
+ *          description: Internal server error
+ */
+router.get('/overview', authenticate, UserController.getAccountOverview);
```
Comment (andela-hpopoola) [inline server/routes/user.routes.js:452]:
```suggestion
router.get('/account-overview', authenticate, UserController.getAccountOverview);
```

File: server/services/user.service.js [hunk 6]
```diff
@@ -328,3 +330,20 @@ export const removePropertyFromFavorites = async (favorite) => {
     throw new ErrorHandler(httpStatus.BAD_REQUEST, 'Error removing property from favorites', error);
   }
 };
+
+export const getAccountOverview = async (userId) => {
+  let contributionReward = await calculateContributionReward(userId);
```
Comment (andela-hpopoola) [inline server/services/user.service.js:335]: **Code Smell**

Reassigning contributionReward from an array to a number is not a good programming pattern. Lets create different variables to cater for this.

```js
const calculatedContributionReward = await calculateContributionReward(userId);
const contributionReward = calculatedContributionReward.length > 0 ? calculatedContributionReward[0].contributionReward : 0;
```

Please, lets fix this for totalAmountPaid and referralRewards too 🙏

File: test/controllers/user.controller.test.js [hunk 7]
```diff
@@ -1240,4 +1247,119 @@ describe('User Controller', () => {
...
+        await addReferral(referral);
+        await addProperty(property);
+        await addEnquiry(enquiry);
+        await createOffer(offer);
+        await addTransaction(transaction);
+        await acceptOffer({ userId, offerId, signature: 'https://ballers.ng/signature.png' });
+      });
+
+      context('with valid token', () => {
+        it('returns contribution reward of one million', (done) => {
```
Comment (andela-hpopoola) [inline test/controllers/user.controller.test.js:1322]: If we update the number and it returns a contribution reward of 2 million, we would be forced to update the test description which is not necessary.

```suggestion
        it('returns a valid contribution reward', (done) => {
```

File: test/services/user.service.test.js [hunk 8]
```diff
@@ -761,4 +766,54 @@ describe('User Service', () => {
...
+    });
+
+    beforeEach(async () => {
+      await addUser(user);
+      await addUser(vendor);
+    });
+    context('when account has no contribution rewards', () => {
+      it('returns zero as account overview', async () => {
+        const overview = await getAccountOverview(userId);
+        expect(overview.contributionReward).to.eql(0);
```
Comment (andela-hpopoola) [inline test/services/user.service.test.js:801]: Can we add test that referralRewards and totalAmountPaid is valid here too 🙏

File: test/services/user.service.test.js [hunk 9]
```diff
@@ -761,4 +766,54 @@ describe('User Service', () => {
...
+      beforeEach(async () => {
+        await addProperty(property);
+        await addEnquiry(enquiry);
+        await createOffer(offer);
+        await acceptOffer({ userId, offerId, signature: 'https://ballers.ng/signature.png' });
+      });
+
+      it('returns two million account overview', async () => {
+        const overview = await getAccountOverview(userId);
+        expect(overview.contributionReward).to.eql(2000000);
```
Comment (andela-hpopoola) [inline test/services/user.service.test.js:815]: Can we add test that referralRewards and totalAmountPaid is valid here too 🙏

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for making the changes. This looks good to me, only a few nits to ensure that we improve the code quality. I will also need to test on my computer to be sure that we covered all edge cases. Thank you for making this work.

File: test/controllers/user.controller.test.js [hunk 10]
```diff
@@ -1240,4 +1247,119 @@ describe('User Controller', () => {
...
+
+      context('with valid token', () => {
+        it('returns a valid contribution reward', (done) => {
+          request()
+            .get('/api/v1/user/account-overview')
+            .set('authorization', token)
+            .end((err, res) => {
+              expect(res).to.have.status(200);
+              expect(res.body.success).to.be.eql(true);
+              expect(res.body.accountOverview.contributionReward).to.be.eql(1000000);
```
Comment (andela-hpopoola) [inline test/controllers/user.controller.test.js:1329]: Magic numbers are very difficult to understand when reading. This can be improved to something more meaningful like:
```suggestion     expect(res.body.accountOverview.contributionReward).to.be.eql(property.price - offer.totalAmountPayable);
```

Comment (andela-hpopoola) [inline test/controllers/user.controller.test.js:1329]: Magic numbers are very difficult to understand when reading. This can be improved to something more meaningful like:

```suggestion
              expect(res.body.accountOverview.contributionReward).to.be.eql(property.price - offer.totalAmountPayable);
```

File: test/controllers/user.controller.test.js [hunk 11]
```diff
@@ -1240,4 +1247,119 @@ describe('User Controller', () => {
...
+              expect(res.body.accountOverview.totalAmountPaid).to.be.eql(250000);
```
Comment (andela-hpopoola) [inline test/controllers/user.controller.test.js:1330]:
```suggestion
              expect(res.body.accountOverview.totalAmountPaid).to.be.eql(transaction.amount);
```

File: test/controllers/user.controller.test.js [hunk 12]
```diff
@@ -1240,4 +1247,119 @@ describe('User Controller', () => {
...
+              expect(res.body.accountOverview.referralRewards).to.be.eql(50000);
```
Comment (andela-hpopoola) [inline test/controllers/user.controller.test.js:1331]: Can you fix this and other related tests too as suggested above? The problem of using magic numbers is that it would make the test more difficult to update in the future which could make you waste more time debugging the cause of the bug.

File: test/services/user.service.test.js [hunk 13]
```diff
@@ -761,4 +770,76 @@ describe('User Service', () => {
...
+        await addEnquiry(enquiry);
+        await createOffer(offer);
+        await acceptOffer({ userId, offerId, signature: 'https://ballers.ng/signature.png' });
+        await addReferral(referral);
+        await addTransaction(transaction);
+      });
+
+      it('returns two million account overview', async () => {
+        const overview = await getAccountOverview(userId);
+        expect(overview.contributionReward).to.eql(2000000);
```
Comment (andela-hpopoola) [inline test/services/user.service.test.js:839]: Can we get rid of the magic numbers here too 🙏

**Review by andela-hpopoola — APPROVED** [review]
Thank you for working on this. Feel free to merge after getting rid of all the magic numbers.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
I will have to un-approve this as I noticed that the property section is still missing. From the pivotal tracker story, the returned account overview should be like this:

```
accountOverview: {
           properties: [
              {
                  name : "Property A",
                  paid: "150000",
              },
              {
                 ...
              }
           ],
          totalAmountPaid: 1500000,
          referralRewards: 10000,
          contributionReward: 500000,
}
```

Can we add the property section and test that it really exists please?

**Comment by tobieeniafe** [comment]
> I will have to un-approve this as I noticed that the property section is still missing. From the pivotal tracker story, the returned account overview should be like this:
> 
> ```
> accountOverview: {
>            properties: [
>               {
>                   name : "Property A",
>                   paid: "150000",
>               },
>               { 
>                  ...
>               }
>            ],
>           totalAmountPaid: 1500000,
>           referralRewards: 10000,
>           contributionReward: 500000,
> }
> ```
> 
> Can we add the property section and test that it really exists please?

that would be implemented when story #175070906 is complete. As it would use the same service.

**Review by andela-hpopoola — APPROVED** [review]
Thank you for working on this. In case the list of properties takes a longer time to implement, I feel it is safer to merge this and make those changes in a later PR. 

Feel free to merge this but implement the below comment in another PR to speed up our work. Thank you.

https://github.com/Highrachy/ballers-api/pull/52#pullrequestreview-504481806
